acceptancetests/assess_upgrade.py

```diff
@@ -0,0 +1,199 @@
+
+Bootstrap a previous version of juju and then upgrade the controller and models
+to the new one.
+
+  - Spins up a local streams server (requires to be run on lxd or on a machine
+     with an externally routable interface.
+  - Bootstraps using the 'stable' juju then upgrades to the 'devel' version.
+  - Deploys mediawiki and mysql to ensure workloads continue working.
+"""
+
+from __future__ import print_function
+
+import argparse
+import logging
+import sys
+from subprocess import (
+    CalledProcessError,
+)
+
+
+from collections import (
+    namedtuple,
+    )
+from utility import (
+    add_basic_testing_arguments,
+    configure_logging,
+    temp_dir,
+    )
+from jujupy.binaries import (
+    get_stable_juju
+    )
+from jujupy.client import (
+    ModelClient,
+    get_stripped_version_number,
+)
+from jujupy.stream_server import (
+    StreamServer,
+    agent_tgz_from_juju_binary,
+    )
+from jujupy.wait_condition import (
+    ModelCheckFailed,
+    WaitModelVersion,
+    wait_for_model_check,
+)
+from jujupy.workloads import (
+    deploy_mediawiki_with_db,
+    assert_media_wiki_is_responding
+    )
+
+from utility import (
+    JujuAssertionError
+)
+
+__metaclass__ = type
+
+
+log = logging.getLogger("assess_upgrade")
+
+
+VersionParts = namedtuple('VersionParts', ['version', 'series', 'arch'])
+
+
+def assess_upgrade_from_stable_to_develop(args, stable_bsm, devel_client):
+    stable_client = stable_bsm.client
+    with temp_dir() as base_dir:
+        stream_server = StreamServer(base_dir)
+        setup_agent_metadata(
+            stream_server, args.stable_agent,
+            stable_client, base_dir, 'proposed')
+        setup_agent_metadata(
+            stream_server, args.devel_agent,
+            devel_client, base_dir, 'proposed')
+        with stream_server.server() as url:
+            stable_client.env.update_config({
+                'agent-metadata-url': url,
+                'agent-stream': 'proposed'
+            })
+            with stable_bsm.booted_context(False):
+                assert_stable_model_is_correct(stable_client)
+
+                deploy_mediawiki_with_db(stable_client)
+                assert_media_wiki_is_responding(stable_client)
+                upgrade_stable_to_devel_version(devel_client)
+                assert_media_wiki_is_responding(devel_client)
+
+
+def upgrade_stable_to_devel_version(client):
+    devel_version = get_stripped_version_number(client.version)
+    client.get_controller_client().juju('upgrade-juju', ('-m', 'controller'))
+    assert_model_is_version(client.get_controller_client(), devel_version)
+    wait_until_model_upgrades(client)
+
+    client.juju('upgrade-juju', ())
+    assert_model_is_version(client, devel_version)
+    wait_until_model_upgrades(client)
+
+
+def wait_until_model_upgrades(client, timeout=300):
+    # Poll using a command that will fail until the upgrade is complete.
+    def model_upgrade_status_check(client):
```

Comment:
This feels useful enough to not warrant it to be nested. Thought?

Reply:
I've moved ```wait_until_model_upgrades``` to ```wait_conditions``` so it's generally available.
I'm not sure the nested function is useful outside of the parent ```wait_until_model_upgrades``` as any time you'll want to check the upgrade is successful you'll most likely want to wait for a timeout until it is (and if not, set ```timeout``` to 0 or 1.